### DISCUSSION TITLE
How to import a targets file multiple times with different parameters?

### DISCUSSION THREAD

**Discussion by Nefcanto** [dotnet/msbuild#6835] [Q&A]
I have an API that imports many modules. Here's the code snippet:

```
<Project  Sdk="Microsoft.NET.Sdk.Web">
    // rest of the MSBuild code in Api.csproj
    <Import Project="/CRM/Ticketing/Module.targets" />
    <Import Project="/CRM/Customers/Module.targets" />
    <Import Project="/CRM/Loyalty/Module.targets" />
    <Import Project="/CRM/Club/Module.targets" />
<Project>
```

And each of those `Module.targets` files define two properties and import a base `Module.targets` file:
***
/CRM/Customers/Module.targets
```
<Project>
    <PropertyGroup>
        <Project>CRM</Project>
        <Module>Customers</Module>
    </PropertyGroup>
    
    <Import Project="/Base/Module.targets" />
</Project>
```
***
/CRM/Club/Module.targets
***
```
<Project>
    <PropertyGroup>
        <Project>CRM</Project>
        <Module>Club</Module>
    </PropertyGroup>
    
    <Import Project="/Base/Module.targets" />
</Project>
```
***
And this is the base `Module.targets`:

```
<Project>
  
  <ItemGroup>
    <Reference Include="$(Project).$(Module).Models">
      <HintPath>/$(Project)/$(Module)/Packages/$(Project).$(Module).Models.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="$(Project).$(Module).DataAccess">
      <HintPath>/$(Project)/$(Module)/Packages/$(Project).$(Module).DataAccess.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="$(Project).$(Module).Business">
      <HintPath>/$(Project)/$(Module)/Packages/$(Project).$(Module).Business.dll</HintPath>
      <Private>True</Private>
    </Reference>
  </ItemGroup>
  
  <ItemGroup>
    <Content Include="/$(Project)/$(Module)/Database.json">
      <Link>Database\$(Module).json</Link>
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
  </ItemGroup>

</Project>
```

So, basically what I'm trying to do is to centralize a module's MSBuild configuration and only provide parameters.

The problem is that MSBuild complains that:

> warning MSB4011: "/Base/Module.targets" cannot be imported again. It was already imported at "/CRM/Customers/Project.targets (8,3)". This is most likely a build authoring error. This subsequent import will be ignored. [/CRM/CustomerApi/Api.csproj]

How can I solve this?

**Comment by rainersigwald** [comment]
This isn't possible in MSBuild's import model. The only way to run code again with "parameters" is by specifying properties to the MSBuild task, which doesn't seem highly appropriate here. You might consider generating the files from a template so you can expand the boilerplate-like things?

**Reply by Nefcanto** [reply]
Can you give a quick simple example of your last line? Template stuff I mean.

**Reply by mikerochip** [reply]
Sorry for the necro, Google search brought me here because I was also looking for a way to parameterize and re-use MSBuild commands. These blog posts might help?

* https://www.jrdodds.com/blog/2021/8/15/msbuild-about-import-guards
* https://www.jrdodds.com/blog/2021/8/8/how-the-use-of-calltarget-can-be-a-code-smell-in-msbuild

Basically: use conditional import guards, and use `DependsOn`
